Improve InvokeAI cross attention reliability and speed when using MPS for large images

brkirch opened the pull request:
The InvokeAI cross attention optimization will cause the MPS backend to crash if it attempts to use a `slice_size` divisible by 4096 (which is the case when generating a resolution of 1024x1024, 2048x512, 512x2048, 1024x512, 512x1024) or an even batch size:
```
  0%|                                                    | 0/20 [00:00<?, ?it/s]/AppleInternal/Library/BuildRoots/5b8a32f9-5db2-11ed-8aeb-7ef33c48bc85/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShaders/MPSCore/Types/MPSNDArray.mm:724: failed assertion `[MPSNDArray initWithDevice:descriptor:] Error: total bytes of NDArray > 2**32'
./webui.sh: line 169:  2011 Abort trap: 6           "${python_cmd}" "${LAUNCH_SCRIPT}" "$@"
```
This PR uses the lower memory InvokeAI MPS cross attention optimization when q.shape[0] * q.shape[1] >= 2**18 (which is the case with 1024x1024 resolution or larger or batch sizes of 4 or greater) or q.shape[0] % 32 == 0 (even batch size) because in my testing it has slightly better speed in such cases as well as not crashing. 1024x512 and 512x1024 still have the potential to cause a crash, so to prevent that the `slice_size` is reduced by 1 if it is divisble by 4096, which should usually allow for image generation to proceed normally.

rworne commented:
I tried this out and I'm still getting a crash with 1024x1024 and 1600x1600:

```
To create a public link, set `share=True` in `launch()`.
100%|█████████████████████████████████████████████████████████████████████████| 20/20 [00:19<00:00,  1.04it/s]
Total progress: 100%|█████████████████████████████████████████████████████████| 20/20 [00:18<00:00,  1.08it/s]
100%|█████████████████████████████████████████████████████████████████████████| 20/20 [00:18<00:00,  1.07it/s]
  0%|                                                                                  | 0/20 [00:00<?, ?it/s]/AppleInternal/Library/BuildRoots/5b8a32f9-5db2-11ed-8aeb-7ef33c48bc85/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShaders/MPSCore/Types/MPSNDArray.mm:724: failed assertion `[MPSNDArray initWithDevice:descriptor:] Error: total bytes of NDArray > 2**32'
./webui.sh: line 169: 11400 Abort trap: 6           "${python_cmd}" "${LAUNCH_SCRIPT}" "$@"
Robert@Mac-Studio stable-diffusion-webui % /opt/homebrew/Cellar/python@3.10/3.10.8/Frameworks/Python.framework/Versions/3.10/lib/python3.10/multiprocessing/resource_tracker.py:224: UserWarning: resource_tracker: There appear to be 1 leaked semaphore objects to clean up at shutdown
  warnings.warn('resource_tracker: There appear to be %d '
```

So it's possible there may be more than one location that needs to be fixed?

brkirch commented:
With further testing I'm still seeing some issues, so I'll do some more work on this to see if I can find a better solution.

brkirch commented:
@rworne What version of macOS are you using and how much RAM does your computer have?

rworne commented:
> @rworne What version of macOS are you using and how much RAM does your computer have?

Ventura 13.1, 32GB (Base model Mac Studio).

brkirch committed:
Use other MPS optimization for large q.shape[0] * q.shape[1]

Check if q.shape[0] * q.shape[1] is 2**18 or larger and use the lower memory usage MPS optimization if it is. This should prevent most crashes that were occurring at certain resolutions (e.g. 1024x1024, 2048x512, 512x2048).

Also included is a change to check slice_size and prevent it from being divisible by 4096 which also results in a crash. Otherwise a crash can occur at 1024x512 or 512x1024 resolution.

brkirch commented:
@rworne Thank you for the information. I just made some changes, please give them a try. More specifically, try generating images with resolutions of 1024x1024 and 1024x512. 1600x1600 still won't work and this is expected; only generating images with less than 2^21 (2,097,152) pixels seems to work.

rworne commented:
> @rworne Thank you for the information. I just made some changes, please give them a try. More specifically, try generating images with resolutions of 1024x1024 and 1024x512. 1600x1600 still won't work and this is expected; only generating images with less than 2^21 (2,097,152) pixels seems to work.

Tried 1024x1024 and 1024x512 (with highresfix) and all seems well.

However the bug is still tickled when you run "batch size" with an even number with the width and height of 512x512.

brkirch commented:
> However the bug is still tickled when you run "batch size" with an even number with the width and height of 512x512.

I am unable to reproduce that, but the changes currently in this PR shouldn't in any way directly affect 512x512 image generation so I'm going to mark this PR as ready to review for now. I'll continue to look into what could be wrong with batch generation and put in another PR if I can figure out a fix for it.

rworne commented:
> I am unable to reproduce that, but the changes currently in this PR shouldn't in any way directly affect 512x512 image generation so I'm going to mark this PR as ready to review for now. I'll continue to look into what could be wrong with batch generation and put in another PR if I can figure out a fix for it.

For the purposes of this fix, this is good enough and the batch issue can be another bug to fix later.

To trigger the batch issue, all I have to do is generate anything with an initial size of 512x512, set the batch size to 2 or 4 (1024 or 2048) sounds familiar?  And I get the same error and crash we see in this issue. 3 works fine.  
256x256 I cannot get it to crash at any number up to the max (8).

rworne commented:
@brkirch:
Found another way to make it crash.  It seems to work OK with cpkt files, but when using safetensors it again will not allow 512x1024.  I think this is how it was crashing when I was using the batch function.

brkirch commented:
@rworne Sorry, it turns out I was changing the wrong batch setting (`Batch count`, for some reason I overlooked `Batch size`). I was able to reproduce the issue and I've made some modifications accordingly, see if it works correctly for you now.

rworne commented:
> @rworne Sorry, it turns out I was changing the wrong batch setting (`Batch count`, for some reason I overlooked `Batch size`). I was able to reproduce the issue and I've made some modifications accordingly, see if it works correctly for you now.

Just tested it out, looks good - even batch works now for even numbers.

rachelcenter commented:
comfyUi says "reconnecting..." in the middle of certain video2video workflows, is that because the m2 ultra mac 128gb computer cant handle it?

when it says "reconnecting" on screen, terminal on mac says:

0%| | 0/12 [00:00<?, ?it/s]/AppleInternal/Library/BuildRoots/91a344b1-f985-11ee-b563-fe8bc7981bff/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShaders/MPSCore/Types/MPSNDArray.mm:788: failed assertion `[MPSNDArray initWithDevice:descriptor:] Error: total bytes of NDArray > 2**32'

/Users/rachel/.pyenv/versions/3.12.4/lib/python3.12/multiprocessing/resource_tracker.py:254: UserWarning: resource_tracker: There appear to be 1 leaked semaphore objects to clean up at shutdown

warnings.warn('resource_tracker: There appear to be %d '

zsh: abort python3.12 [main.py](http://main.py/)

rachelcenter commented:
> > @rworne Sorry, it turns out I was changing the wrong batch setting (`Batch count`, for some reason I overlooked `Batch size`). I was able to reproduce the issue and I've made some modifications accordingly, see if it works correctly for you now.
> 
> Just tested it out, looks good - even batch works now for even numbers.

can you tell me how to fix it? I'm new to coding.
